### ISSUE TITLE
Hp drain kills despite prevention

### ISSUE THREAD

**Issue by ZeitHeld** [ShadowMario/FNF-PsychEngine#11518]
### Describe your problem here.

Okay, I used a script to modify the health you gain, and have that combined with a health drain script. even a prevention, not do die from it, if you hp is too low.
I tested it on spookeez, and somehow... if both sing toegther, I receive the loss, and die-
Is there any way to fix this? Or anything that can like, prevent it?
(And before that, I had to figure out, that I have to drain the og health at least, because otherwise, it would add ontop again-)

The active code that is used:
```
function onCreate()
	--hpgain = 0.023
	hpgain = 0
	hploss = 0.0475
	malus = 0.023
	--malus can be modified if you like!
end

function goodNoteHit(id, noteData, noteType, isSustainNote)
	--debugPrint('before-hp:', getProperty('health'))

	if isSustainNote then
		setProperty('health', health + hpgain/2)
		--debugPrint('sus ', gain/2)
	else
		setProperty('health', health + hpgain)
		--debugPrint(gain)
	end

	--debugPrint('mid-hp:', getProperty('health'))

	health = getProperty('health')

	setProperty('health', health - 0.023)
        --0.023 is default health gain value
	--also, substracting, while it otherwise would stack ontop

	-debugPrint('end-hp:', getProperty('health'))
	--debugPrint('### ##### ###')
end

function noteMiss(id, noteData, noteType, isSustainNote)
	health = getProperty('health')

	setProperty('health', health + 0.0475)
	--0.0475 is default health loss value
	--also, adding, while it otherwise would stack ontop

	if isSustainNote then
		setProperty('health', health - hploss/2)
	else
		setProperty('health', health - hploss)
	end
end

function opponentNoteHit(id, noteData, noteType, isSustainNote)
	health = getProperty('health')

	if health > malus then
		if isSustainNote then
			setProperty('health', health - malus/2)
		else
			setProperty('health', health - malus)
		end
	end
end
```

### Are you modding a build from source or with Lua?

Lua

### What is your build target?

Windows x64

### Did you edit anything in this build? If so, mention or summarize your changes.

No

**Comment by DisIsNOt** [comment]
idk may be add a check if the hp was to low or smth

**Comment by ZeitHeld** [comment]
I already did that check on the opponentnotehit()

Before it was even said to 0.1, which is quite low already, but I still
died.

AshMostly ***@***.***> schrieb am So., 18. Dez. 2022, 12:33:

**Comment by DisIsNOt** [comment]
![image](https://user-images.githubusercontent.com/78010753/208363202-e7772757-3cdf-4f9f-bce8-6795de1a522e.png)


may be try this on goodNotehit section

**Comment by ZeitHeld** [comment]
But I like need to always reduce the hp, because otherwise I'd gain hp.
The health gets added AFTER the lua function. But yes... that could be the
issue. Maybe? Even though the other function prevents it...

AshMostly ***@***.***> schrieb am Mo., 19. Dez. 2022, 07:38:

**Comment by Draconic-Hybrid** [comment]
If your drain is larger than the limit, it will still kill. Take this for example.

Your limit to stop draining is 2. (Using whole numbers for easier explanation) However, the drain is 5. What happens if the current HP is 3?

Well, the HP is greater than 2, so it'll drain, leaving you with -2 as a result, which kills you.

Either lower the drain or raise the limit.

**Comment by ZeitHeld** [comment]
that is why I check, if the remaining hp are greater than the drain.
but I think, at this point, the functions happen almost simultaneously, so that lua can't go really after it, and just does the yeet

**Comment by Draconic-Hybrid** [comment]
I think I misread your code a bit. You want hits to not do anything, but opponents hitting the notes to drain?

**Comment by ZeitHeld** [comment]
Misses do need to drain. But on the opponentNoteHit it should only drain
If currentHealth > hpDrain
Because then it usually does drain, to prevent killing off, if the player
doesn't get their turn after a few notes. Otherwise, there would be
unpassable sections

Draconic-Hybrid ***@***.***> schrieb am Mi., 4. Jan. 2023,
01:22:

**Comment by Draconic-Hybrid** [comment]
Yeah, I'm a bit unsure about this. In theory, it shouldn't be killing you. Maybe two things are happening at the same time to kill you, or it's a different script causing it entirely.

**Comment by ZeitHeld** [comment]
ain't another script for sure. so it's something weird, with function calling. even though it shouldn't be able to-
